File: check_i18n.js
const fs = require('fs');
const dir = './src/web/ui/js/';
const src = fs.readFileSync(dir + 'i18n.js', 'utf8');
const start = src.indexOf('const I18N = {');
const bodyStart = src.indexOf('{', start);
function parseLangs(text, from) {
  let i = from + 1; const langs = {};
  while (i < text.length) {
    while (i < text.length && /[\s,]/.test(text[i])) i++;
    if (text[i] === '}') break;
    let m = /^['"]?([\w-]+)['"]?\s*:\s*\{/.exec(text.slice(i));
    if (!m) break;
    const name = m[1]; let j = i + m[0].length - 1; let depth = 1; j++;
    const objStart = j;
    while (j < text.length && depth > 0) { if (text[j] === '{') depth++; else if (text[j] === '}') depth--; j++; }
    const objBody = text.slice(objStart, j - 1);
    const keys = new Set(); const re = /['"]([\w.{}\- ]+?)['"]\s*:/g; let km;
    while ((km = re.exec(objBody))) keys.add(km[1]);
    langs[name] = keys; i = j;
  }
  return langs;
}
const langs = parseLangs(src, bodyStart);
const names = Object.keys(langs);
console.log('Languages:', names.map(n => n + '(' + langs[n].size + ')').join(', '));
const en = langs.en; let ok = true;
for (const n of names) {
  if (n === 'en') continue;
  const missing = [...en].filter(k => !langs[n].has(k));
  const extra = [...langs[n]].filter(k => !en.has(k));
  if (missing.length) { ok = false; console.log(n, 'MISSING', missing); }
  if (extra.length) { ok = false; console.log(n, 'EXTRA', extra); }
}
const files = fs.readdirSync(dir).filter(f => f.endsWith('.js'));
const used = new Set();
for (const f of files) { const c = fs.readFileSync(dir + f, 'utf8'); const re = /\btr\(\s*['"]([\w.\-]+?)['"]/g; let m; while ((m = re.exec(c))) used.add(m[1]); }
const missingKeys = [...used].filter(k => !en.has(k) && k !== 'theme.');
console.log('Used tr() keys:', used.size);
if (missingKeys.length) { ok = false; console.log('MISSING in dict:', missingKeys.sort()); }
else console.log('All used tr() keys exist (ignoring dynamic theme.).');
console.log(ok ? 'CONSISTENCY_OK' : 'CONSISTENCY_FAIL');

File: scripts/check_i18n.py
import re, os
d = 'src/web/ui/js/'
src = open(d + 'i18n.js', encoding='utf-8').read()
start = src.index('const I18N = {')
i = src.index('{', start) + 1
langs = {}
while i < len(src):
    while i < len(src) and src[i] in ' \t\r\n,':
        i += 1
    if src[i] == '}':
        break
    m = re.match(r"['\"]?([\w-]+)['\"]?\s*:\s*\{", src[i:])
    if not m:
        break
    name = m.group(1); j = i + m.end() - 1; depth = 1; j += 1; s0 = j
    while j < len(src) and depth > 0:
        if src[j] == '{': depth += 1
        elif src[j] == '}': depth -= 1
        j += 1
    langs[name] = re.findall(r"['\"]([\w.{}\- ]+?)['\"]\s*:", src[s0:j-1])
    i = j
names = list(langs)
print('Languages:', ', '.join(f'{n}({len(set(langs[n]))})' for n in names))
en = set(langs['en']); ok = True
for n in names:
    if n == 'en': continue
    if set(en) - set(langs[n]): ok = False; print(n, 'MISSING', sorted(set(en) - set(langs[n])))
    if set(langs[n]) - en: ok = False; print(n, 'EXTRA', sorted(set(langs[n]) - en))
for n in names:
    seen = {}
    for k in langs[n]: seen[k] = seen.get(k, 0) + 1
    dd = [k for k, c in seen.items() if c > 1]
    if dd: ok = False; print(n, 'DUP', dd)
used = set()
for f in os.listdir(d):
    if f.endswith('.js'):
        used |= set(re.findall(r"\btr\(\s*['\"]([\w.\-]+?)['\"]", open(d + f, encoding='utf-8').read()))
miss = [k for k in used if k not in en and k != 'theme.']
if miss: ok = False; print('USED-MISSING', sorted(miss))
print('OK' if ok else 'FAIL')
